Add echo checkup

orelmisan opened the pull request:
This adds a basic checkup that shall be used as:
1. A debugging tool for developing the project.
2. A reference to create more advanced and useful checkups.

The checkup basically receives a string as an input, and writes it without modification to it's output.

This adds an option to make.sh to unit test the checkup's entrypoint script:
```bash
./automation/make.sh --echo-unit-test
```

This adds an option to make.sh to build the checkup's image:
```bash
./automation/make.sh --echo-build-image
```
It supports podman as a default, with the ability to choose arbitrary container engines such as Docker:
```bash
CRI=docker ./automation/make.sh --echo-build-image
```
An image tag can also be set:
```bash
ECHO_IMAGE_TAG=0.1.0 ./automation/make.sh --echo-build-image
```
Signed-off-by: Orel Misan <omisan@redhat.com>

EdDev commented on example-checkups/echo/Dockerfile:
```diff
@@ -0,0 +1,11 @@
+FROM alpine:3.15.0
```
For documentation purposes, this is fine.
But once this is to be executed, it cannot point to dockerhub due to their rate limiting access.

orelmisan replied:
Changed to `ubi-minimal:8.5-240.1648458092`.

EdDev commented on example-checkups/echo/entrypoint:
```diff
@@ -0,0 +1,77 @@
+
+echo "*** Starting echo checkup ***"
+
+# Check mandatory environment variables supplied by the framework
```
Please split these into functions and see if their names can drop the comments.

orelmisan replied:
Done

EdDev reviewed (changes requested):
This is an important functionality that we are introducing.
Please, cover it by tests, locally and on the CI.

Regarding its location in the directory tree, what do you think about this structure:
```
root
+-example
     |
     +-checkup
         |
         +echo
```

EdDev commented:
You intend to publish the container?

orelmisan commented:
IMO we should create a repository under https://quay.io/ for this project and put the approved images there.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/752430999c405bcf4dfa4d1c4a6c25ce4fdad67a..2d58e13a62145217bd30fb32edd939412590ce75): Create a new directory hierarchy as suggested by @EdDev [here](https://github.com/kiagnose/kiagnose/pull/6#pullrequestreview-911599765).

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/2d58e13a62145217bd30fb32edd939412590ce75..2870e2a7a520f041f927e3b5a8fba17917458b35): Changed base image to `registry.access.redhat.com/ubi8/ubi-minimal:8.5-240.1648458092`

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/2870e2a7a520f041f927e3b5a8fba17917458b35..26da918539cc088d4cc1f7931368dda92845e18d): Added unit tests.

EdDev commented on examples/checkups/echo/entrypoint_test:
```diff
@@ -0,0 +1,134 @@
+entrypoint_should_succeed_with_all_env_vars() {
+  echo "entrypoint should exit(0) with all required env variables"
+
+  RESULT_CONFIGMAP_NAMESPACE=ns1 \
```
Please use local variables, otherwise these pollute the global space.

orelmisan replied:
In this specific case, the variables are affecting only the command that comes after them:
```
MESSAGE="Hi" ./entrypoint
```

Only entrypoint "sees" the variable.

EdDev commented:
I am missing where you run the tests written in this change.

ormergi commented on examples/checkups/echo/entrypoint:
```diff
@@ -0,0 +1,101 @@
+}
+
+create_output_file() {
+  output_file="$1"
```
Please reduce the scope of the variables inside functions and set them read-only where possible (e.g: `local -r path="$1"`) (all over)

orelmisan replied:
Done.

ormergi commented on examples/checkups/echo/entrypoint:
```diff
@@ -0,0 +1,101 @@
+KUBECTL=${KUBECTL:-kubectl}
+OUTPUT_FILE="/tmp/result_configmap.yaml"
+
+check_required_env_vars_from_framework() {
```
Please add the `function` keyword to functions definitions (all over)

orelmisan replied:
This is a matter of style.
The suggested style of function definition is coming from [ksh](https://en.wikipedia.org/wiki/KornShell), and is not part of the [POSIX standard](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_09_05).
More info about this subject can be found [here](https://unix.stackexchange.com/questions/73750/difference-between-function-foo-and-foo).

bash supports them [both](https://www.gnu.org/software/bash/manual/html_node/Shell-Functions.html).

ormergi replied:
As far as I know, we dont ahve to be strict about keeping the scripts 100% POSIX compatible especially where we
are going to provide the container image, we will make sure it contains the correct shell for this script to work.

It will defiantly make the code more readable as it may look closer to more structured programming languages
similar to what to the the tests.

EdDev replied:
I do not see the benefit really, but if you feel strong about it, it is an easy change you could suggest.
I am ok with both options.

ormergi replied:
I am not insisting on it, we can do this change later.

ormergi commented on examples/checkups/echo/entrypoint:
```diff
@@ -0,0 +1,101 @@
+  print_all_inputs
+  create_output_file "${OUTPUT_FILE}"
+
+  echo "Trying to patch the result ConfigMap object..."
+  ${KUBECTL} patch configmap "${RESULT_CONFIGMAP_NAME}" -n "${RESULT_CONFIGMAP_NAMESPACE}" --patch-file "${OUTPUT_FILE}"
```
Please add error handling so that if the patch is failing we could get informative error.

ormergi commented on examples/checkups/echo/entrypoint:
```diff
@@ -0,0 +1,101 @@
+
+check_required_env_vars_from_framework() {
+  if [ -z "$RESULT_CONFIGMAP_NAMESPACE" ]; then
+    echo "$0 error: RESULT_CONFIGMAP_NAMESPACE is empty" >&2
```
Why do we care for "$0"?

orelmisan replied:
This is done in order to distinguish the output coming from this script file.

orelmisan replied:
Removed it.

ormergi commented on examples/checkups/echo/entrypoint:
```diff
@@ -0,0 +1,101 @@
+  namespace: $RESULT_CONFIGMAP_NAMESPACE
+data:
+  status.succeeded: "true"
+  status.failureReason: none
```
Note that `status.failureReason` should remain an empty string (`""`) to prevent inconsistent 
state of this field, please change it accordingly.

orelmisan replied:
This is defined in the checkup author's API [here](https://github.com/kiagnose/kiagnose/pull/11/files#diff-6b36f30998c1570886f749e687f826bd24c8b3fa94415973040d2b9ade26dd2cR328).

orelmisan replied:
Done here (but not in the document).

ormergi commented on examples/checkups/echo/entrypoint:
```diff
@@ -0,0 +1,101 @@
+  fi
+
+  if ! check_required_env_vars_by_echo_checkup; then
+    exit 2
```
Why there is a need to return different error codes? isn't retcode 1 is sufficient?

orelmisan replied:
1 is for fatal errors (We don't know where the ConfigMap is).
2 is for non fatal error - we don't have any data to write to the ConfigMap, only status.

ormergi replied:
It seems that one retcode could enough, since the reason for failure (fatal or not fatal) can be realized from stderr,
for me its confusing, it would be more simple and easier to understand if there was single retcode.

Nevertheless I am not insisting on doing this change now.

ormergi commented on examples/checkups/echo/entrypoint:
```diff
@@ -0,0 +1,101 @@
+set -e
+
+KUBECTL=${KUBECTL:-kubectl}
+OUTPUT_FILE="/tmp/result_configmap.yaml"
```
Please consider renaming it to be more informative, something like `results_configmap_manifest`

orelmisan replied:
Done.

ormergi reviewed (changes requested):
Hi, overall looks good.
Please see my comments.

orelmisan commented on examples/checkups/echo/manifests/dev/echo-checkup-workspace.yaml:
```diff
@@ -0,0 +1,43 @@
+---
```
This is an example for how the framework shall run the checkup.
It sets up the same K8s objects in the same manner as the framework would.

orelmisan replied:
The filename was changed.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/26da918539cc088d4cc1f7931368dda92845e18d..dc0c09a7b02f91cc0103b594c77717cb1c0e8810): Rebase

orelmisan commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/dc0c09a7b02f91cc0103b594c77717cb1c0e8810..b72f94dcd25f06188ce1b0d7ec3ec6d0c84d168f): Added a flag to make.sh to build the echo-checkup's image.
Updated the README file and development Job manifest accordingly.

EdDev commented:
@orelmisan , what do you think about removing the `example` folder and just have the `checkup` one?
It feels a bit odd that we place them in examples and then release them. WDYT?

orelmisan commented:
> @orelmisan , what do you think about removing the `example` folder and just have the `checkup` one? It feels a bit odd that we place them in examples and then release them. WDYT?

No problem.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/b72f94dcd25f06188ce1b0d7ec3ec6d0c84d168f..6e4f53fd885b08222d8ae2431a06d8e587eb9300): Moved checkups/ directory one level up.

orelmisan commented on examples/checkups/echo/entrypoint:
```diff
@@ -0,0 +1,101 @@
+  fi
+
+  print_all_inputs
+  create_output_file "${OUTPUT_FILE}"
```
We can do it, the downside of it is adding logic to the `fake_kubectl` script.
Can this be postponed to a future PR?

EdDev replied:
Yes, this is not a blocker for sure. But we should recognize it is super ugly to use temporary files.

orelmisan commented on examples/checkups/echo/entrypoint:
```diff
@@ -0,0 +1,101 @@
+  status.succeeded: "true"
+  status.failureReason: none
+  status.result.echo: "$MESSAGE"
+...
```
In YAML three dots ("...") indicate the end of a document without starting a new one.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/6e4f53fd885b08222d8ae2431a06d8e587eb9300..e3eb28d7fcf1f0686d14e9f5f89049f6008889a4): Moved the Dockerfile back to the echo/ directory.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/e3eb28d7fcf1f0686d14e9f5f89049f6008889a4..047caa6da97cd8ec42264835d92239bf607f29df): Unified the manifest to run this checkup without the framework.

orelmisan commented on examples/checkups/echo/entrypoint:
```diff
@@ -0,0 +1,101 @@
+  output_file="$1"
+
+  cat <<EOF > "${output_file}"
+---
```
As said in another comment, this is done in order to create the manifest file for kubectl to apply.
This is written to a file (and not to kubectl's stdin) in order to test it using a simple `diff` command.

orelmisan commented on examples/checkups/echo/README.md:
```diff
@@ -0,0 +1,47 @@
+3. "MESSAGE" - a message to write to the results ConfigMap object.
+
+## Outputs
+The following key-value pairs shall be written the `data` field of the results ConfigMap object:
```
Done.

orelmisan commented on examples/checkups/echo/README.md:
```diff
@@ -0,0 +1,47 @@
+- You have "Admin" permissions on the K8s cluster.
+- kubectl is configured to connect to your cluster.
+### Steps
+1. Deploy the checkup required workspace using:
```
Done.

orelmisan commented on examples/checkups/echo/README.md:
```diff
@@ -0,0 +1,47 @@
+1. Deploy the checkup required workspace using:
+
+   `kubectl create -f manifests/dev/echo-checkup-workspace.yaml`
+2. Edit the image tag in `manifests/dev/echo-checkup-job.yaml`.
```
Removed this line.

orelmisan commented on examples/checkups/echo/entrypoint_test:
```diff
@@ -0,0 +1,134 @@
+}
+main() {
+  if ! entrypoint_should_succeed_with_all_env_vars; then
+    exit 1
```
Done.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/568cce3722e081a786c2363a92d2edb4bc26039c..fe3c84973c17fa52932a89f7d4707b7d6cdcbfc1): Updated the README file.

EdDev reviewed (commented):
> echo-checkup: Add manifest to run the checkup manually 

I am not clear what is the expected results from running this.
Could you add info about what one would expect to see?

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/fe3c84973c17fa52932a89f7d4707b7d6cdcbfc1..17a5f835a2cda4c10062f8361c806959215f559f): Added --unit-test-echo option to make.sh

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/17a5f835a2cda4c10062f8361c806959215f559f..91856c575642fa7c667bbab3e6728e65aa51a831): Added CI integration for echo checkup's unit test.
Changed the echo flags to --echo-unit-test and --echo-build-image

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/91856c575642fa7c667bbab3e6728e65aa51a831..6e1f67289f19b8f30e994ba8662383e5e65d09e3): Updated commit message to reflect the change to --echo-build-image.

orelmisan commented:
Change: Updated README  to reflect the change to --echo-build-image.

EdDev reviewed (dismissed):
Looks great, thanks!

@ormergi , please check if the responses satisfy you.

ormergi reviewed (approved):
There were few changes I asked to do,but it seems they are not blockers and we can do them later.
Thanks for the changes, it looks good.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/c4c4a33358bdc00cee7439376410630bf2a43fb1..edd907ab5f9b35142a47896b5ff49fd07e74cc17): Rebase
